source/Plugins.cpp

```diff
@@ -345,6 +507,132 @@ const Set<Plugin> &Plugins::Get()
 // Toggles enabling or disabling a plugin for the next game restart.
 void Plugins::TogglePlugin(const string &name)
 {
+	lock_guard<mutex> guard(pluginsMutex);
 	auto *plugin = plugins.Get(name);
 	plugin->currentState = !plugin->currentState;
 }
+
+
+
+future<void> Plugins::Install(InstallData *installData, bool update)
+{
+	oldNetworkActivity = true;
+	if(!update)
+	{
+		lock_guard<mutex> guard(activePluginsMutex);
+		if(!activePlugins.insert(installData->name).second)
+			return future<void>();
+	}
+
+	return async(launch::async, [installData, update]() noexcept -> void
+		{
+			string zipLocation = Files::Plugins() + installData->name + ".zip";
+			bool success = Download(installData->url, zipLocation);
+			if(success)
+			{
+				success = ExtractZIP(
```

Comment:
Before extracting, is there a way to check if the user has enough free space?

Reply:
now that we have filesystem, yes

Reply:
actually, no because I only know the size of the zip after expanding